### DISCUSSION TITLE
Attribute on complex-type is missing in type-generation

### DISCUSSION THREAD

**Discussion by orbdex** [phpro/soap-client#420] [Q&A] [answered]
Hey there, 

this is a great package! Thanks for maintaining.
We're using this toolset to build a SOAP-Interface to our SAP Business ByDesign based on multiple WSDL definitions from SAP. Works finde but I've encountered a problem with one of the complex types. 

The WSDL definds the following complex type:

            <xsd:complexType name="Quantity">
                <xsd:annotation>
                    <xsd:documentation xml:lang="EN">
                        <ccts:RepresentationTerm>Quantity</ccts:RepresentationTerm>
                    </xsd:documentation>
                </xsd:annotation>
                <xsd:simpleContent>
                    <xsd:extension base="Quantity.Content">
                        <xsd:attribute name="unitCode" type="MeasureUnitCode"/>
                    </xsd:extension>
                </xsd:simpleContent>
            </xsd:complexType>
            <xsd:simpleType name="Quantity.Content">
                <xsd:restriction base="xsd:decimal">
                    <xsd:totalDigits value="31"/>
                    <xsd:fractionDigits value="14"/>
                    <xsd:maxInclusive value="99999999999999999.99999999999999"/>
                    <xsd:minInclusive value="-99999999999999999.99999999999999"/>
                </xsd:restriction>
            </xsd:simpleType>

This type is referenced in another complex-Type (PurchaseRequestItemManually):
`<xsd:element name="Quantity" type="xi4:Quantity" minOccurs="0"/>`

When running the type-generation from CLI the type-class-files are created but the "unitCode"-Attribute is missing in the Type "PurchaseRequestItemManually" for the Quantity-element. It just sets it as a simple float-Value.

I think it's a configuration issues but I don't get the hack. Any ideas?

From a SOAP-UI-Project I extracted the "target XML" we need, the relevant part is:
`<Quantity unitCode="HUR"> 10 </Quantity>`

I can set the parameter to 10 (float) but I need to change the unitCode. 
Would appreciate any ideas and hints.

Thanks guys!

**Comment by veewee** [comment]
Hello,

Thanks for reporting

Is it possible to share the WSDL or create a WSDL that resembles the issue?
What does the generation config look like?
What does the generated code look like?


It might also help if you share the signature of the type from `SoapClient->__getTypes()`, because it is possible that that information just isn't available ATM.

**Comment by orbdex** [comment]
Hey there,

thanks a lot for the reply!
Find the sample WSDL attached (used .txt ending due to limitations here) and the generated types within the zip file.

[sample_wsdl.txt](https://github.com/phpro/soap-client/files/8087893/sample_wsdl.txt)
[Type.zip](https://github.com/phpro/soap-client/files/8087919/Type.zip)

The SOAP-Client Config looks like this:
```php
<?php

use Phpro\SoapClient\CodeGenerator\Assembler;
use Phpro\SoapClient\CodeGenerator\Config\Config;
use Phpro\SoapClient\CodeGenerator\Rules;
use Phpro\SoapClient\Soap\Driver\ExtSoap\ExtSoapEngineFactory;
use Phpro\SoapClient\Soap\Driver\ExtSoap\ExtSoapOptions;

return Config::create()
    ->setEngine($engine = ExtSoapEngineFactory::fromOptions(
        ExtSoapOptions::defaults('src/Lib/ApiClients/SAPByD/Wsdl/ManagePurchaseRequestIn.wsdl', [])
            ->disableWsdlCache()
    ))
    ->setTypeDestination('src/Lib/ApiClients/SAPByD/ManagePurchaseRequestIn/Type')
    ->setTypeNamespace('App\Lib\ApiClients\SAPByD\ManagePurchaseRequestIn\Type')
    ->setClientDestination('src/Lib/ApiClients/SAPByD/ManagePurchaseRequestIn')
    ->setClientName('SAPByDManagePurchaseRequestInClient')
    ->setClientNamespace('App\Lib\ApiClients\SAPByD\ManagePurchaseRequestIn')
    ->setClassMapDestination('src/Lib/ApiClients/SAPByD/ManagePurchaseRequestIn')
    ->setClassMapName('SAPByDManagePurchaseRequestInClassmap')
    ->setClassMapNamespace('App\Lib\ApiClients\SAPByD\ManagePurchaseRequestIn')
    ->addRule(new Rules\AssembleRule(new Assembler\GetterAssembler(new Assembler\GetterAssemblerOptions())))
    ->addRule(new Rules\AssembleRule(new Assembler\ImmutableSetterAssembler(
        new Assembler\ImmutableSetterAssemblerOptions()
    )))
    ->addRule(
        new Rules\IsRequestRule(
            $engine->getMetadata(),
            new Rules\MultiRule([
                new Rules\AssembleRule(new Assembler\RequestAssembler()),
                new Rules\AssembleRule(new Assembler\ConstructorAssembler(new Assembler\ConstructorAssemblerOptions())),
            ])
        )
    )
    ->addRule(
        new Rules\IsResultRule(
            $engine->getMetadata(),
            new Rules\MultiRule([
                new Rules\AssembleRule(new Assembler\ResultAssembler()),
            ])
        )
    )
;
```

**Reply by orbdex** [reply]
One note: I removed the URLs of our SAP-system endpoint in the WSDL and replaced the host-information by xxxxxxxxx

**Comment by veewee** [comment, marked as answer]
It does create this class:

```php

class Quantity
{
    /**
     * @var \App\Lib\ApiClients\SAPByD\ManagePurchaseRequestIn\Type\QuantityContent
     */
    private $_;

    /**
     * @var string
     */
    private $unitCode;
}
```

The `_` here is the (float) value of the xml element and the `unitCode` is the attribute.
You normally can set this to PurchaseRequestItemManually->quantity (but strict types won't work)

**Reply by orbdex** [reply]
Thanks a lot...! Looks good, I did not know the `_` - property. Maybe should have read the documentation better ;-) thanks!

**Reply by veewee** [reply]
It's a "hidden" feature of php's ext-soap 👯

**Reply by orbdex** [reply]
A nice feature ;-) 

I was a bit confused because the phpDoc of the parent-element PurchaseRequestItemManually defined the parameter as float. I now created the Quantity-entity and when I pass this to the withQuantity-function it worked. Thanks again.

`    /**
     * @param float $Quantity
     * @return PurchaseRequestItemManually
     */
    public function withQuantity($Quantity)
    {
        $new = clone $this;
        $new->Quantity = $Quantity;

        return $new;
    }`

**Reply by veewee** [reply]
Not sure why it typehints it as float ATM and if we can improve that - will have to look into that.
